Adds core UI framework types and UI event handler for second core

cosparks opened the pull request:
## UI Framework
### new types
- **MenuViewExample**: example menu view to demonstrate how to use ui framework classes to build a simple menu
- **VisualElement**: base class for all ui elements
- **UIElement**: extends **VisualElement** and is base class for all interactive ui elements
- **Container**: extends **VisualElement** and is base class for all container types--provides means of automatically formatting the elements that it contains
- **UIButton**: basic interactive ui type which provides styling for focus and selection
- **TextComponent**: handles font type, colour and size for any ui type which deals with text
- **TextElement**: ui type for static text
- **UIEventHandler**: interface from main core to UI core--allows main application to hand events and animations over to be run on the second core (events are function pointers, lambdas or *closures*)

### UI-Related
- **main.cpp**: added example of adding 
- **ui/ui_util.h**: provides a number of useful types and functions related to ui
- **graphics/bitmaps.h**: for monochrome bitmap images
- **graphics/colour.h**: for aggregating 16-bit 5-6-5 RGB colour macros
- **graphics/fonts.h**: for aggregating font macros

### Unrelated changes and fixes
- adds static beginAll method to Handleable
- renames old Button class to HardwareButton to distinguish between ui and io button types

cosparks committed:
adds skeletons for ui classes

cosparks committed:
Modifications to ui classes
* replaces VisualElement constructor parameters with setters
* renames DisplayWindow to Container
* adds Adafruit_ILI9314 as param to ui constructors

cosparks committed:
Completes skeletons for core UI and Visual element classes
- UIButton: encapsulates ui button behavior -- focus/unfocus/selection
- TextElement: encapsulates static ui text
- TextComponent: encapsulates bitmap font handling; owned by VisualElement

cosparks committed:
Adds TODOs and new supermileage logo bitmap

cosparks committed:
Expands upon core UI types and adds example
- implements draw methods for VisualElement, Container, TextElement and TextComponent
- implements Container::align method -- currently only aligns elements to center
- adds example ui menu class for experimentation and to demonstrate the usage of ui types

cosparks committed:
fixes build issue caused by git's case insensitivity

cosparks committed:
removes UI folder (already replaced with ui)

cosparks committed:
Fixes compiler warning
- comparison between signed and unsigned type in Handleable.cpp

cosparks committed:
Removes virtual declaration from VisualElement::clear
- base class knows enough about position/size/colour

cosparks committed:
Merge branch 'dev-ui' of github.com:supermileage/dynamometer-firmware into dev-ui

cosparks committed:
rename var in VisualElement::clear

cosparks committed:
Replaces Adafruit_ILI9341 with Adafruit_GFX in core ui classes
- ui framework shouldn't use concrete type if it's not necessary

cosparks committed:
Implements focus and unfocus methods for UIElement
- adds methods for setting background and font focus colours
- adds demo code for focus/unfocus in MenuExample.cpp

cosparks committed:
Adds core ui event handler
- provides api for running animations and arbitrary events on second core (ui core)
- uses mutexes for safe access of event queue
- adds function library from std so users can make use of closures and fn pointers
- updates menu example to make use of UIEventHandler

cosparks committed:
Adds abstract animation class with brief description

cosparks committed:
fixes typo in animation

cosparks committed:
replaces hard-coded paths with env vars in .vscode/launch.json

cosparks committed:
Adds TODOs for alignment in Container.cpp

cosparks committed:
removes .vscode/launch.json and adds it to .gitignore

cosparks committed:
remove DataLogger class

LaPommeCosmique commented on src/ui/VisualElement.h:
```diff
@@ -0,0 +1,98 @@
+         * 
+         * @note width/height match _width/_height fields by default; max width = 320 and max height = 240
+        */
+        VisualElement& addBorder(uint16_t colour, int16_t width = -1, int16_t height = -1);
```
what is height in this context?

LaPommeCosmique replied:
ah I think i maybe understand.. is it something like this?
![image](https://github.com/supermileage/dynamometer-firmware/assets/48578176/046a6034-456d-4e0b-a815-4a3be287e7fa)

cosparks replied:
I left a comment below on how borders are interpreted.  You could could achieve ^ this with nested containers if you wanted.

LaPommeCosmique replied:
hmm I'm still little confused, could you point me to the comment please? I cant seem to find it

LaPommeCosmique commented on src/ui/VisualElement.cpp:
```diff
@@ -0,0 +1,82 @@
+    _display.fillRect(_position.x, _position.y, _width, _height, colour);
+}
+
+void VisualElement::_drawInternal(uint16_t colour) {
```
hmm, should we be drawing the box at the back (the border) before the box at the front?

cosparks replied:
```Adafruit_GFX::fillRect``` fills a rectangle, but ```Adafruit_GFX::drawRect``` only draws the outline, so the border is an outline of a rectangle.

cosparks replied:
Assumption is that the border will sit inside the bounds of its visual element.  I've considered adding a borderThickness field for borders--we could draw extra lines to create a thicker border--but this didn't strike me as something that needed to be done immediately.

LaPommeCosmique replied:
ah i see, that makes sense

LaPommeCosmique commented on src/ui/VisualElement.cpp:
```diff
@@ -0,0 +1,82 @@
+    } else {
+        colour = _backgroundColour;
+    }
+    _display.fillRect(_position.x, _position.y, _width, _height, colour);
```
does this account for the border?

cosparks replied:
yes

LaPommeCosmique commented on src/ui/Container.h:
```diff
@@ -0,0 +1,90 @@
+        Container& addVisualElements(std::vector<VisualElement*>& elements);
+
+        /**
+         * @brief add set of visual elements contained within window display
```
add ~~set of~~ visual element~~s~~

LaPommeCosmique commented on src/ui/Container.cpp:
What are your thoughts on adding the a transparent background option for the Container and TextElement class?

cosparks replied:
As it is, visual elements will only draw their background if it is a different colour from their parent's, so transparency can be achieved in that way.

LaPommeCosmique commented on src/ui/TextComponent.cpp:
```diff
@@ -0,0 +1,69 @@
+}
+
+void TextComponent::clear(Adafruit_GFX& display) {
+    _drawInternal(display, _owner->getBackgroundColour());
```
Note: I'm not sure if this is an issue, because I'm not sure when ```clear()``` will be called

However, if we have a UIButton, that is currently in a focus/active state, ```getBackgroundColour()``` would not return the actual color of the element. 

I wonder if it is worth adding something like ```getCurrentColour()``` function for this reason

cosparks replied:
```VisualElement::clear``` will most likely only be used for transitioning between frames in animations (picture a blinking animation that indicates that the BPM is on).

Good catch though.  ```TextComponent::clear``` should be also be called by TextElement and UIButton and it is not

cosparks replied:
After looking at this again, it makes sense to me that it isn't called anywhere just yet because there aren't any examples of text animations.  I've added a colour parameter to ```TextComponent::clear``` for controllability, but this is definitely one of those things that might have to change depending on how it's used

LaPommeCosmique commented on src/ui/Container.cpp:
If I understand correctly, it is possible to have the child elements in a container overflow past of the bounds of the container. If this happens, the inherited ```clear()``` function is not adequate to fully remove all the visual elements of the container

LaPommeCosmique replied:
I wonder if its better redrawing the entire display instead of calling ```clear()``` when we want to remove an element, because I feel like there are many edge cases that might cause weird behaviour (for example, if an element overlaps another element that is not its parent element).

cosparks replied:
To me, it doesn't make sense for child elements to have larger dimensions than their parents.  We could auto-resize containers if that is the case, but it's also trivial for a developer to make sure that child elements are the same size or smaller than their parents.

silviu-toderita reviewed (approved):
This seems really cool and comprehensive. To be honest, I didn't go through all the code, and it seems a bit tricky to do so without working on it directly. But it seems easy to understand, the structure is clear, and I feel like I could build a UI pretty quickly by diving into the code. 

Out of curiosity, what is the "simpler form of styling" that you proposed on discord? Is it using a different language, or changing the structure of it in C++?

silviu-toderita commented on src/application/MenuViewExample.cpp:
```diff
@@ -0,0 +1,120 @@
+    _window->draw();
+}
+
+void MenuViewExample::run() {
```
What is happening here exactly? Is this something to do with interactivity? And why is there a ```delay(1000)``` on line 115?

cosparks replied:
This is just lazy demo code that walks through all the selectable elements of the example menu, focusing and unfocusing each one.  The delay is there to leave some time between focus / unfocus so you can see what it looks like on the display.

cosparks commented:
> Out of curiosity, what is the "simpler form of styling" that you proposed on discord? Is it using a different language, or changing the structure of it in C++?

@silviu-toderita 

I completely understand why html and xml are used for styling UIs now.  I wasn't thinking of going that far, but I was thinking of adding config structs for each of the visual element classes:

```
struct VisualElementConfig {
    uint16_t backgroundColour;
    int16_t width;
    int16_t height;
    ...
};
```

Then you could just define config macros for all the generic styled elements which would be consistent throughout the whole project.

In a more recent branch, I've actually gone a different route though and instead added a style.h file with size/colour/font macros for all generic elements (like default menu-button styling and whatnot), and an app_util file with functions that configure the elements.

cosparks committed:
Changes for pr
- modifies TextComponent::clear -> overloaded TextComponent::draw with colour parameter
- fixes comments on Container::addVisualElement

cosparks commented:
>In a more recent branch, I've actually gone a different route though and instead added a style.h file with size/colour/font macros for all generic elements (like default menu-button styling and whatnot), and an app_util file with functions that configure the elements.

@silviu-toderita 

This is what I mean: https://github.com/supermileage/dynamometer-firmware/pull/38
